docs/docs/adrs/adr-015-partial-set-security.md

```diff
@@ -0,0 +1,293 @@
+
+```
+ChainIdAndConsAddrKey(ValidatorOptedOutBytePrefix, chainID, addr.ToSdkConsAddr())
+```
+
+We augment the state of the provider to keep track for each chain all the validators that have opted in.
+```
+// returns all the validators that have opted out on chain `chainID`
+func (k Keeper) GetOptedOutValidators(ctx sdk.Context, chainID string) []validators
+```
+
+We also update the state of the opted-in validators when a validator has opted out. We do this check in every `BeginBlock`.
+
+We introduce a query to retrieve the validators that are about to be opted out. The query would operate as follows.
+```
+gaiad query provider optedOutValidators $chainId
+```
+
+### When does a consumer chain start?
+A Top N consumer chain starts at the specified date (`spawn_time`) if the [`ConsumerAdditionProposal`](https://github.com/cosmos/interchain-security/blob/v3.3.0/proto/interchain_security/ccv/provider/v1/provider.proto#L27) has passed. An Opt In consumer chain starts if at least one validator has opted in. We check this in [BeginBlockInit](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/proposal.go#L372):
+```
+func (k Keeper) BeginBlockInit(ctx sdk.Context) {
+    propsToExecute := k.GetConsumerAdditionPropsToExecute(ctx)
+
+    for _, prop := range propsToExecute {
+        chainId := prop.ChainId
+        if !k.IsTopN(ctx, chainId) && len(k.GetOptedInValidators(ctx, chainId)) == 0 {
+            // drop the proposal
+            ctx.Logger().Info("could not start chain because noone has opted in")
+            continue
+        }   
+        ...
+```
+
+### How do we send the partial validator sets to the consumer chains?
+To send the validator set to a consumer chain we first have to generate the validator updates with a function similar to this one:
+
+```
+func (k Keeper) GetPartialSetValidators(ctx types.Context, chainId string valUpdates []abci.ValidatorUpdate) (newUpdates []abci.ValidatorUpdate)
+
+    powerSum := sdk.ZeroDec()
+    totalPower := k.ComputeTotalPower(ctx)
+    topNThreshold := k.GetTopN(ctx, chainId) / 100
+    for _, val := range valUpdates {
+        powerSum = powerSum.Add(sdk.NewDecFromInt(sdk.NewInt(val.Power)))
+        // if powerSum / totalPower > topNThreshold
+        if k.IsTopN(ctx, chainId) && powerSum.Quo(totalPower).GT(topNThreshold) {
+            // is Top N validator
+            updates = append(updates, abci.ValidatorUpdate{
+                PubKey: val.PubKey,
+                Power:  val.Power,
+            })
+       } else {
+       if k.IsValidatorOptedIn(ctx, chainId, val.PubKey) {
+            updates = append(updates, abci.ValidatorUpdate{
+                PubKey: val.PubKey,
+                Power:  val.Power,
+            })
+        } else {
+            // remove this validator
+            updates = append(updates, abci.ValidatorUpdate{
```

Comment:
Nice catch! There ways to keep track of the validator set that the consumer chain is consuming and not send unnecessary validator updates. I'll try to think more on if there's an easy way to do so but won't investigate too much time on this because when we move to [epochs](https://github.com/cosmos/interchain-security/blob/main/docs/docs/adrs/adr-014-epochs.md) this would unlikely be an issue.

Reply:
Intend to revamp the whole section here. But in regards to your question:
> Is it fine for Comet to receive a val update with power 0 for a validator that is not part of the set?

It does not seem to be fine. From what I see, in CometBFT, when we "delete"(i.e., set power to 0) more validators than the actual validators of the chain we get a [panic](https://github.com/cometbft/cometbft/blob/v0.38.5/types/validator_set.go#L550). Something like this could trigger, if in a consumer chain with only 5 opted in validators, we send 175 deleted ones.

Reply:
Resolving this since we removed the partial-set computation from the ADR for now.